### DISCUSSION TITLE
API响应时间阈值

### DISCUSSION THREAD

**Discussion by dogourd** [HXSecurity/DongTai-agent-java#394] [Q&A] [answered]
[fb64e2d2: fix: api fallback](https://github.com/HXSecurity/DongTai-agent-java/commit/fb64e2d2ab2791a3994eda01314ad1f1775e20fb#diff-cb439bc953cd61758f94da20b560a3c641ae35de651596405a1d4eebaed78e10)  
GraphBuilder 中静态的 URL 和 URI 似乎是线程不安全的，URL 看起来会作为后续请求是否有必要对HttpServletRequest及HttpServletResponse对象包装的依据，是我没发现对应的同步代码？

**Comment by lostsnow** [comment, marked as answer]
There is no relationship between `URL`/`URI` in `GraphBuilder` and `HttpServletRequest/HttpServletResponse`, only when the circuit breaker will be used to, the circuit breaker will be re-implemented and will not use this again

**Comment by dogourd** [comment]
What you said doesn't seem to match the code. Look at here:
`io.dongtai.iast.core.handler.hookpoint.SpyDispatcherImpl#leaveHttp`
```java
if (EngineManager.SCOPE_TRACKER.isExitedHttp() && EngineManager.isEnterHttp()) {
                    EngineManager.maintainRequestCount();
                    GraphBuilder.buildAndReport(request, response);
                    EngineManager.cleanThreadState();
                    long responseTimeEnd = System.currentTimeMillis()-responseTime.get()+8;
                    DongTaiLog.debug("url {} response time: {} ms",GraphBuilder.getURL(),responseTimeEnd);
                    if (RemoteConfigUtils.enableAutoFallback() && responseTimeEnd > RemoteConfigUtils.getApiResponseTime(null)){
                        RemoteConfigUtils.fallbackReqCount++;
                        DongTaiLog.warn("url {} response time: {} ms, greater than {} ms",GraphBuilder.getURL(),responseTimeEnd,RemoteConfigUtils.getApiResponseTime(null));
                        if (!"/".equals(GraphBuilder.getURL())){
                            ConfigMatcher.getInstance().FALLBACK_URL.add(GraphBuilder.getURI());
                        }
                    }
                }
```
The URI is added to FALLBACK_URL by `ConfigMatcher.getInstance().FALLBACK_URL.add(GraphBuilder.getURI());`. Then look at the HttpImpl class.
```java
public static Object cloneResponse(Object response, boolean isJakarta) {
        try {
            if (response == null) {
                return null;
            }
            if (ConfigMatcher.getInstance().disableExtension((String) REQUEST_META.get().get("requestURI"))) {
                return response;
            }
            if (ConfigMatcher.getInstance().getBlackUrl(REQUEST_META.get())) {
                return response;
            }
            if (cloneResponseMethod == null) {
                loadCloneResponseMethod();
            }
            return cloneResponseMethod.invoke(null, response, isJakarta);
        } catch (IllegalAccessException e) {
            return response;
        } catch (InvocationTargetException e) {
            return response;
        } finally {
            REQUEST_META.remove();
        }
    }
```
`if (ConfigMatcher.getInstance().getBlackUrl(REQUEST_META.get())) {
                return response;
            }`
The preceding FALLBACK URL is used to determine if the response object needs to be wrapped.
Some http responses may miss wrapping if there are concurrency issues causing unexpected urls to be added to the set.

**Comment by dogourd** [comment]
My fault, it's no big deal if some packaging is missed

**Comment by lostsnow** [comment]
As mentioned above, this is only used for circuit breaker, for skipping some requests with long response times, affecting at most a fraction of the requests collected, and may replace this dynamic blacklist in the future by remote configuration
